crates/trie/sparse/src/trie.rs

```diff
@@ -261,6 +261,125 @@ impl<P: BlindedProvider> SparseTrie<P> {
     }
 }
 
+/// A representation for nibbles, that uses an even/odd flag.
+#[derive(Clone, PartialEq, Eq, Debug, Hash, Default)]
+pub struct PackedNibbles {
+    /// The nibbles themselves, stored as a byte array.
+    pub nibbles: SmallVec<[u8; 32]>,
+    /// The even/odd flag, indicating whether the length is even or odd.
+    pub even: bool,
+}
+
+impl PackedNibbles {
+    /// Returns whether or not the PackedNibbles is empty.
+    pub fn is_empty(&self) -> bool {
+        self.nibbles.is_empty()
+    }
+
+    /// Extends this PackedNibbles with the given PackedNibbles.
+    fn extend_path(&mut self, other: &PackedNibbles) {
+        // if we're odd we need to extend starting from the next nibble
+        //
+        // TODO: figure out if we should have it in the low or high nibble
+        // for example if we have 0x123, we're storing it as [0x12, 0x30], and we want to extend the
+        // path with 0x456, we need to end up with [0x12, 0x34, 0x56]
+        //
+        // this involves shifting everything over one nibble and merging the first byte of the
+        // second path with with the last byte of the first path
+        //
+        // If we're even, we're fine, and can just extend.
+        if other.is_empty() {
+            // if the other is empty, we can just return
+            return
+        }
+
+        if self.nibbles.is_empty() {
+            // if we're empty, we can just use the other nibbles
+            self.nibbles.copy_from_slice(&other.nibbles);
+            self.even = other.even;
+            return
+        }
+
+        // whether even or odd, we need to extend the nibbles
+        let prev_len = self.nibbles.len();
+        self.nibbles.extend_from_slice(&other.nibbles);
```

Comment:
this should probably happen only in self.even case, else do loop one by one shifting it in

Reply:
makes sense